flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/MultipleInputStreamTask.java

```diff
@@ -175,10 +176,23 @@ protected void createInputProcessor(
                         getEnvironment().getTaskInfo());
     }
 
+    protected Optional<CheckpointBarrierHandler> getCheckpointBarrierHandler() {
+        return Optional.ofNullable(checkpointBarrierHandler);
+    }
+
     @Override
     public Future<Boolean> triggerCheckpointAsync(
             CheckpointMetaData metadata, CheckpointOptions options) {
 
+        if (operatorChain.getSourceTaskInputs().size() == 0) {
+            return super.triggerCheckpointAsync(metadata, options);
+        }
+
+        // If there are chained sources, we would always only trigger
+        // the chained sources for checkpoint. This means that for
+        // the checkpoints during the upstream task finished and
+        // this task receives the EndOfPartitionEvent, the checkpoint
+        // would not subsume the pending ones.
```

Comment:
Is that comment correct? We will still subsume the pending checkpoints from within the sources. We do process barriers for the source inputs in that case.

Reply:
Yes, we should indeed also be able to subsume the checkpoints in this case, and the actual difference is that we would not insert barriers for the unfinished network inputs and it needs to wait for the EndOfPartitionEvent. I'll fix the comments~